File: examples/tiny-gizmo-example/gizmo_dragger.cpp
#include "gizmo_dragger.h"
#include <assert.h>
#include <iostream>

void TRSGizmo::begin(const Vector2 &cursor) {
  float best_t = std::numeric_limits<float>::infinity();
  for (auto &target : this->_scene) {
    //
    // ray intersection
    //
    switch (this->_visible) {
    case GizmoMode::Translation: {
      auto intersection = tinygizmo::TranslationGizmo::intersect(
          _current_state, _local_toggle, target->transform);
      if (intersection) {
        auto [ray_state, active_component] = *intersection;
        if (ray_state.t < best_t) {
          best_t = ray_state.t;
          this->_t = active_component;
          this->_gizmo_target = target;
          this->_ray_state = ray_state;
        }
      }
      break;
    }

    case GizmoMode::Rotation: {
      auto intersection = tinygizmo::RotationGizmo::intersect(
          _current_state, _local_toggle, target->transform);
      if (intersection) {
        auto [ray_state, active_component] = *intersection;
        if (ray_state.t < best_t) {
          best_t = ray_state.t;
          this->_r = active_component;
          this->_gizmo_target = target;
          this->_ray_state = ray_state;
        }
      }
      break;
    }

    case GizmoMode::Scaling: {
      auto intersection = tinygizmo::ScalingGizmo::intersect(
          _current_state, _local_toggle, target->transform, _uniform);
      if (intersection) {
        auto [ray_state, active_component] = *intersection;
        if (ray_state.t < best_t) {
          best_t = ray_state.t;
          this->_s = active_component;
          this->_gizmo_target = target;
          this->_ray_state = ray_state;
        }
      }
      break;
    }

    default:
      assert(false);
      throw std::runtime_error("unknown gizmo mode");
    }
  }
}

void TRSGizmo::end(const Vector2 &cursor) {
  _t = {};
  _r = {};
  _s = {};
  _gizmo_target = {};
  _ray_state = {};
}

void TRSGizmo::drag(const DragState &state, int w, int h,
                    const Vector2 &cursor) {
  if (auto target = _gizmo_target) {
    switch (_visible) {
    case GizmoMode::Translation:
      if (_t) {
        target->transform = tinygizmo::TranslationGizmo::drag(
            *_t, _current_state, _ray_state, target->transform);
      }
      break;
    case GizmoMode::Rotation:
      if (_r) {
        target->transform = tinygizmo::RotationGizmo::drag(
            *_r, _current_state, _ray_state, target->transform);
      }
      break;
    case GizmoMode::Scaling:
      if (_s) {
        target->transform = tinygizmo::ScalingGizmo::drag(
            *_s, _current_state, _ray_state, target->transform);
      }
      break;
    }
  }
}

void TRSGizmo::hotkey(int w, int h, const Vector2 &cursor,
                      const struct Hotkey &hotkey) {
  if (hotkey.hotkey_ctrl == true) {
    if (_last_hotkey.hotkey_translate == false &&
        hotkey.hotkey_translate == true) {
      _visible = GizmoMode::Translation;
    }
    if (_last_hotkey.hotkey_rotate == false && hotkey.hotkey_rotate == true) {
      _visible = GizmoMode::Rotation;
    }
    if (_last_hotkey.hotkey_scale == false && hotkey.hotkey_scale == true) {
      _visible = GizmoMode::Scaling;
    }

    if (hotkey.hotkey_local) {
      _local_toggle = !_local_toggle;
      std::cout << "_local_toggle: " << _local_toggle << std::endl;
    }
  }
  _last_hotkey = _current_hotkey;
  _current_hotkey = hotkey;

  auto ray = GetMouseRay(cursor, *_camera);
  auto rot =
      QuaternionFromEuler(ray.direction.x, ray.direction.y, ray.direction.z);

  _last_state = _current_state;
  _current_state = {
      .mouse_down = IsMouseButtonDown(MOUSE_BUTTON_LEFT),
      // optional flag to draw the gizmos at a constant screen-space
      // scale gizmo_state.screenspace_scale = 80.f; camera projection
      .viewport_size = {static_cast<float>(w), static_cast<float>(h)},
      .ray =
          {
              .origin = {ray.position.x, ray.position.y, ray.position.z},
              .direction = {ray.direction.x, ray.direction.y, ray.direction.z},
          },
      .cam_yfov = 1.0f,
      .cam_orientation = {rot.x, rot.y, rot.z, rot.w},
  };
}

void TRSGizmo::load(Drawable *drawable) {
  _positions.clear();
  _colors.clear();
  _indices.clear();
  tinygizmo::AddTriangleFunc add_world_triangle =
      [self = this](const tinygizmo::Float4 &rgba, const tinygizmo::Float3 &p0,
                    const tinygizmo::Float3 &p1, const tinygizmo::Float3 &p2) {
        //
        auto offset = self->_positions.size();
        Color color{
            static_cast<unsigned char>(std::max(0.0f, rgba.x) * 255),
            static_cast<unsigned char>(std::max(0.0f, rgba.y) * 255),
            static_cast<unsigned char>(std::max(0.0f, rgba.z) * 255),
            static_cast<unsigned char>(std::max(0.0f, rgba.w) * 255),
        };
        self->_positions.push_back({p0.x, p0.y, p0.z});
        self->_positions.push_back({p1.x, p1.y, p1.z});
        self->_positions.push_back({p2.x, p2.y, p2.z});
        self->_colors.push_back(color);
        self->_colors.push_back(color);
        self->_colors.push_back(color);
        self->_indices.push_back(offset + 0);
        self->_indices.push_back(offset + 1);
        self->_indices.push_back(offset + 2);
      };

  for (auto &target : this->_scene) {
    auto [draw_scale, p, ray] = _current_state.gizmo_transform_and_local_ray(
        _local_toggle, target->transform);
    auto gizmoMatrix = p.matrix() * tinygizmo::Float4x4::scaling(
                                        draw_scale, draw_scale, draw_scale);
    switch (this->_visible) {
    case GizmoMode::Translation:
      tinygizmo::TranslationGizmo::mesh(gizmoMatrix, add_world_triangle, _t);
      break;
    case GizmoMode::Rotation:
      tinygizmo::RotationGizmo::mesh(gizmoMatrix, add_world_triangle, _r);
      break;
    case GizmoMode::Scaling:
      tinygizmo::ScalingGizmo::mesh(gizmoMatrix, add_world_triangle, _s);
      break;
    }
  }

  if (_positions.size() && _indices.size()) {
    drawable->load(_positions.size(), _positions.data(), _colors.data(),
                   _indices.size(), _indices.data(), true);
  }
}

File: examples/tiny-gizmo-example/gizmo_dragger.zig
const std = @import("std");
const c = @import("c.zig");
const drawable = @import("drawable.zig");
const rdrag = @import("rdrag.zig");
const tinygizmo = @import("tinygizmo");

pub const GizmoMode = enum {
    Translation,
    Rotation,
    Scaling,
};

pub const GizmoActiveComponent = union(GizmoMode) {
    Translation: ?tinygizmo.TranslationGizmo.GizmoComponentType,
    Rotation: ?tinygizmo.RotationGizmo.GizmoComponentType,
    Scaling: ?tinygizmo.ScalingGizmo.GizmoComponentType,
};

pub const Hotkey = struct {
    hotkey_ctrl: bool = false,
    hotkey_translate: bool = false,
    hotkey_rotate: bool = false,
    hotkey_scale: bool = false,
    hotkey_local: bool = false,
};

pub const TRSGizmo = struct {
    mode: GizmoActiveComponent = .{ .Translation = null },
    camera: *c.Camera,
    scene: []*drawable.Drawable,
    gizmo_target: ?*drawable.Drawable = null,

    current_hotkey: Hotkey = .{},
    last_hotkey: Hotkey = .{},
    current_state: tinygizmo.FrameState = .{},
    last_state: tinygizmo.FrameState = .{},
    local_toggle: bool = true,
    uniform: bool = true,

    ray_state: tinygizmo.RayState = .{},

    positions: std.ArrayList(c.Vector3),
    colors: std.ArrayList(c.Color),
    indices: std.ArrayList(c_ushort),

    pub fn init(
        allocator: std.mem.Allocator,
        camera: *c.Camera,
        scene: []*drawable.Drawable,
    ) !@This() {
        return .{
            .positions = std.ArrayList(c.Vector3).init(allocator),
            .colors = std.ArrayList(c.Color).init(allocator),
            .indices = std.ArrayList(c_ushort).init(allocator),
            .camera = camera,
            .scene = scene,
        };
    }

    pub fn deinit(self: *@This()) void {
        self.positions.deinit();
        self.colors.deinit();
        self.indices.deinit();
    }

    pub fn begin(self: *@This(), _: c.Vector2) void {
        var best_t = std.math.inf(f32);
        for (self.scene) |target| {
            //
            // ray intersection
            //
            switch (self.mode) {
                .Translation => {
                    if (tinygizmo.TranslationGizmo.intersect(
                        self.current_state,
                        self.local_toggle,
                        target.transform,
                    )) |intersection| {
                        const ray_state, const active_component = intersection;
                        if (ray_state.t < best_t) {
                            best_t = ray_state.t;
                            self.mode = .{ .Translation = active_component };
                            self.gizmo_target = target;
                            self.ray_state = ray_state;
                        }
                    }
                },

                .Rotation => {
                    if (tinygizmo.RotationGizmo.intersect(
                        self.current_state,
                        self.local_toggle,
                        target.transform,
                    )) |intersection| {
                        const ray_state, const active_component = intersection;
                        if (ray_state.t < best_t) {
                            best_t = ray_state.t;
                            self.mode = .{ .Rotation = active_component };
                            self.gizmo_target = target;
                            self.ray_state = ray_state;
                        }
                    }
                },

                .Scaling => {
                    if (tinygizmo.ScalingGizmo.intersect(
                        self.current_state,
                        self.local_toggle,
                        target.transform,
                        self.uniform,
                    )) |intersection| {
                        const ray_state, const active_component = intersection;
                        if (ray_state.t < best_t) {
                            best_t = ray_state.t;
                            self.mode = .{ .Scaling = active_component };
                            self.gizmo_target = target;
                            self.ray_state = ray_state;
                        }
                    }
                },
            }
        }
    }

    pub fn end(self: *@This(), _: c.Vector2) void {
        self.mode = switch (self.mode) {
            .Translation => .{ .Translation = null },
            .Rotation => .{ .Rotation = null },
            .Scaling => .{ .Scaling = null },
        };
        self.gizmo_target = null;
        self.ray_state = .{};
    }

    pub fn drag(
        self: *@This(),
        _: rdrag.DragState,
        _: i32,
        _: i32,
        _: c.Vector2,
    ) void {
        if (self.gizmo_target) |target| {
            switch (self.mode) {
                .Translation => |t| {
                    if (t) |active_component| {
                        target.transform = tinygizmo.TranslationGizmo.drag(
                            active_component,
                            self.current_state,
                            self.ray_state,
                            target.transform,
                        );
                    }
                },

                .Rotation => |r| {
                    if (r) |active_component| {
                        target.transform = tinygizmo.RotationGizmo.drag(
                            active_component,
                            self.current_state,
                            self.ray_state,
                            target.transform,
                        );
                    }
                },

                .Scaling => |s| {
                    if (s) |active_component| {
                        target.transform = tinygizmo.ScalingGizmo.drag(
                            active_component,
                            self.current_state,
                            self.ray_state,
                            target.transform,
                        );
                    }
                },
            }
        }
    }

    pub fn process_hotkey(self: *@This(), w: i32, h: i32, cursor: c.Vector2, hotkey: Hotkey) void {
        if (hotkey.hotkey_ctrl == true) {
            if (self.last_hotkey.hotkey_translate == false and
                hotkey.hotkey_translate == true)
            {
                self.mode = .{ .Translation = null };
            } else if (self.last_hotkey.hotkey_rotate == false and
                hotkey.hotkey_rotate == true)
            {
                self.mode = .{ .Rotation = null };
            } else if (self.last_hotkey.hotkey_scale == false and
                hotkey.hotkey_scale == true)
            {
                self.mode = .{ .Scaling = null };
            }

            self.local_toggle = if (!self.last_hotkey.hotkey_local and hotkey.hotkey_local)
                !self.local_toggle
            else
                self.local_toggle;
        }

        self.last_hotkey = self.current_hotkey;
        self.current_hotkey = hotkey;

        const ray = c.GetMouseRay(cursor, self.camera.*);
        const rot =
            c.QuaternionFromEuler(ray.direction.x, ray.direction.y, ray.direction.z);

        self.last_state = self.current_state;
        self.current_state = .{
            .mouse_down = c.IsMouseButtonDown(c.MOUSE_BUTTON_LEFT),
            // optional flag to draw the gizmos at a constant screen-space
            // scale gizmo_state.screenspace_scale = 80.f; camera projection
            .viewport_size = .{ .x = @floatFromInt(w), .y = @floatFromInt(h) },
            .ray = .{
                .origin = .{
                    .x = ray.position.x,
                    .y = ray.position.y,
                    .z = ray.position.z,
                },
                .direction = .{
                    .x = ray.direction.x,
                    .y = ray.direction.y,
                    .z = ray.direction.z,
                },
            },
            .cam_yfov = 1.0,
            .cam_orientation = .{ .x = rot.x, .y = rot.y, .z = rot.z, .w = rot.w },
        };
    }

    fn add_triangle(
        user: *anyopaque,
        rgba: tinygizmo.Float4,
        p0: tinygizmo.Float3,
        p1: tinygizmo.Float3,
        p2: tinygizmo.Float3,
    ) void {
        const self = @as(*TRSGizmo, @ptrCast(@alignCast(user)));
        const offset: c_ushort = @intCast(self.positions.items.len);
        const color = c.Color{
            .r = @intFromFloat(@max(0.0, rgba.x) * 255),
            .g = @intFromFloat(@max(0.0, rgba.y) * 255),
            .b = @intFromFloat(@max(0.0, rgba.z) * 255),
            .a = @intFromFloat(@max(0.0, rgba.w) * 255),
        };
        self.positions.append(.{ .x = p0.x, .y = p0.y, .z = p0.z }) catch @panic("append");
        self.positions.append(.{ .x = p1.x, .y = p1.y, .z = p1.z }) catch @panic("append");
        self.positions.append(.{ .x = p2.x, .y = p2.y, .z = p2.z }) catch @panic("append");
        self.colors.append(color) catch @panic("append");
        self.colors.append(color) catch @panic("append");
        self.colors.append(color) catch @panic("append");
        self.indices.append(offset + 0) catch @panic("append");
        self.indices.append(offset + 1) catch @panic("append");
        self.indices.append(offset + 2) catch @panic("append");
    }

    pub fn load(self: *@This(), dst: *drawable.Drawable) void {
        self.positions.clearRetainingCapacity();
        self.colors.clearRetainingCapacity();
        self.indices.clearRetainingCapacity();

        for (self.scene) |target| {
            const draw_scale, const p, const ray =
                self.current_state.gizmo_transform_and_local_ray(self.local_toggle, target.transform);
            _ = ray; // autofix

            const scaleMatrix =
                tinygizmo.Float4x4.scaling(draw_scale, draw_scale, draw_scale);
            const gizmoMatrix = p.matrix().mul(scaleMatrix);

            switch (self.mode) {
                .Translation => |_t| {
                    tinygizmo.TranslationGizmo.mesh(gizmoMatrix, self, add_triangle, _t);
                },

                .Rotation => |_r| {
                    tinygizmo.RotationGizmo.mesh(gizmoMatrix, self, add_triangle, _r);
                },

                .Scaling => |_s| {
                    tinygizmo.ScalingGizmo.mesh(gizmoMatrix, self, add_triangle, _s);
                },
            }
        }

        if (self.positions.items.len > 0 and self.indices.items.len > 0) {
            dst.load_slice(
                self.positions.items,
                self.colors.items,
                self.indices.items,
                true,
            );
        }
    }
};
